enhancements/node/lazy-image-pull.md

```diff
@@ -0,0 +1,248 @@
+---
+title: lazy-image-pull
+authors:
+  - @harche
+reviewers: # Include a comment about what domain expertise a reviewer is expected to bring and what area of the enhancement you expect them to focus on. For example: - "@networkguru, for networking aspects, please look at IP bootstrapping aspect"
+  - @mrunalp
+  - @rphillips
+  - @sairameshv
+approvers: # A single approver is preferred, the role of the approver is to raise important questions, help ensure the enhancement receives reviews from all applicable areas/SMEs, and determine when consensus is achieved such that the EP can move forward to implementation.  Having multiple approvers makes it difficult to determine who is responsible for the actual approval.
+  - @mrunalp
+api-approvers: # In case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers). If there is no API change, use "None"
+  - @mrunalp
+creation-date: 2024-03-18
+last-updated: 2024-03-18
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - https://issues.redhat.com/browse/OCPNODE-2204
+---
+
+# Enable Lazy Image Pulling Support in OpenShift
+
+
+## Summary
+
+This enhancement proposes integrating support for lazy image pulling in OpenShift using [Stargz Store plugin](https://github.com/containerd/stargz-snapshotter). This will provide the ability to lazy-pull compatible container images, significantly reducing container startup times and potentially decreasing network strain for the specific workloads that can benefit from it.
+
+## Motivation
+
+* **Faster Serverless:** - Serverless end points need to serve the cold requests as fast as possible. Lazy-pulling of the images allow the serverlesss end point to start serving the user request without getting blocked on the entire container image to get downloaded on the host.
+* **Improved User Experience:** Large images often lead to extended pod startup times. Lazy-pulling drastically reduces the initial delay, providing a smoother user experience.
+* **Potential Benefits for Large Images:** If the container image is shared between various containers which utilize only part of the image during their execution, Lazy image pulling speeds up such container startup by downloading only the relavent bits from those images.
```

Comment:
It would be worth exploring what types of images would not work with stargz. And how we can deduce that?

ie Can we understand why stargz is not working for this images and what kind of logs can we look at to determine that for support/operational?

Reply:
That snapshotter supports `zstd:chunked` and `eStargz` format. OCP will support `zstd:chunked` format only. You can enable debugging in `/etc/container/storage.conf` to get more verbose output of the image handling.

Reply:
That isn't really answering my question.  

> **Potential Benefits for Large Images:** If the container image is shared between various containers which > utilize only part of the image during their execution, Lazy image pulling speeds up such container startup by > downloading only the relavent bits from those images.
this makes it sound like the benefits of lazy pulling is not guaranteed. So i wanted to know how we could know if it wasn't working or working. From a support and a user perspective.

Reply:
Will it always work if someone uses zstd:chunked for their containers. And in all cases, that is all that is needed?

Are there cases where all the bits would need to be downloaded?

Reply:
If the [Stargz service plugin](https://github.com/openshift/enhancements/pull/1600/files#diff-10bfa811cdabc678cbfd947dbeaf04b63ebd24e8d9ab5a05c7abbb998cb72f8aR148) is not running on the host, then the images will not get lazy pulled. i.e. the container would start times will be like regular container images. 

If that plugin is running and the users suspect any defect in the plugin they can look at the unit logs for details. 

You can use image with zstd:chucked format, but without that plugin the lazy part of image pulling won't work, so you would have container start up times comparable to the regular container images.

Reply:
I'm mostly asking because the language you said makes it sound like this isn't guaranteed.

I think if stargz service plugin is enabled, and images will get lazy pulled.

I would maybe just drop the "Potentially". 

But adding some logs on debugging this would be useful for docs/feature enablement.

Reply:
Basically lazy pulling is trading throughput for latency — startup might be faster, but all operations are a bit slower due to the indirection required for the “is the file available already, or do I need to download it” checks.

So, the answer to “what types of images _don’t_ benefit” from lazy pulling is “all well-designed ones” (when throughput matters): if _all files_ in the image are actually necessary, and consumed over the lifetime of the application, then it is faster to pull _and_ to use the image using an ordinary pull, rather than the FUSE slowly-download-files-on-demand lazy approach.

Lazy pulling should be beneficial only if _many_ files in the container are not _actually necessary_ for processing the typical request (and if the files that are necessary are so few that the per-file request latency is not overwhelming). I’m sure that happens frequently in practice, but it is quite likely an indication of something that could be fixed in the image.

---

(There are other benefits to zstd:chunked, like possibly sharing files across container images / layers, but those don’t depend on lazy pulling.)

Reply:
>Basically lazy pulling is trading throughput for latency

stargz-snapshotter will continue to download rest of the image in the background.

Reply:
The FUSE overhead is paid during _every single filesystem operation_ during the container runtime, isn’t it? For a long-running container which is not running purely out of memory (e.g. a single statically linked binary) it will _eventually_ overwhelm any possible savings.